text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,719 @@
+
+```ts
+import Service from '@ember/service';
+
+export default class Session extends Service {
+  login(email: string, password: string) {
+    // ...
+  }
+  
+  logout() {
+    // ...
+  }
+}
+
+declare module '@ember/service' {
+  interface Registry {
+    session: Session;
+  }
+}
+```
+
+Then, users could write fairly idiomatic Ember and have the types resolve correctly:
+
+```ts
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+
+export default Component.extend({
+  session: service('session'), // type: `Session`
+})
+```
+
+The Typed Ember blueprints generate that boilerplate for users, which (hopefully) minimizes the pain of working with these definitions. However, to make this work for the registry APIs, users (and therefore the blueprints) would have to add *another* declaration to the bottom of the file:
+
+```diff
+  import Service from '@ember/service';
+  
+  export default class Session extends Service {
+    login(email: string, password: string) {
+      // ...
+    }
+    
+    logout() {
+      // ...
+    }
+  }
+  
+  declare module '@ember/service' {
+    interface Registry {
+      session: Session;
+    }
+  }
+  
++ declare module 'TBD-some-registry-spot' {
++   interface Registry {
++     'service:session': typeof Session;
++   }
++ }
+```
+
+This *works*, but is additional boilerplate and, more importantly, *duplication of information*, which can get out of sync. Both of these problems are avoided by the new design proposed by this RFC. With the new design, users would still need the same boilerplate required today, but the types for `lookup` and other registry APIs can be extended in a way transparent to users so that they “just work” with *zero* additional effort, as covered in the next section.
+
+### Proposed type definitions
+
+If we were to ship this, we would make one significant change to the *existing* types on DefinitelyTyped, supplying a codemod for ember-cli-typescript users. Where today, the various type registries refer to the class types, we would update them to refer to the *constructor* type for the class. For example, for the service registry:
+
+```diff
+ interface Registry {
+-  session: Session;
++  session: typeof Session;
+ }
+```
+
+This single change unlocks the ability to rigorously and exhaustively supply types for the API proposed above, with no further changes to existing user code.
+
+We would introduce the following types at DefinitelyTyped for the new API, along with rigorous type tests to guarantee they provide the guarantees they specify. (A basic working demo of this is available on [this TypeScript playground][types-demo].)
+
+```ts
+// ----- type utilities ----- //
+
+type ConstructorArgs<T> = T extends new (...args: infer U) => unknown ? U : never;
+
+type InstanceOf<T> = T extends new (...args: any) => infer R ? R : never;
+
+type ConstructorFor<T> = new (...args: ConstructorArgs<T>) => InstanceOf<T>;
```

Comment:
I'm probably missing something subtle (or even completely obvious), but what does type do? At first glance it looks like it would be a no-op if applied to a type with a `new` signature, other than possibly throwing out any potential parametric polymorphism.

Reply:
…huh. Let me play with this. Because now that you mention it, yeah, it should be. 😰

Seems like as long as the *other* constraints are in place in the spots where we use this, and we deploy `typeof` appropriately in the registry, we shouldn't need this, because we'd already have it.

Reply:
I validated that; thank you for catching it. I'll update it later with a change that drops this entirely and makes the type param for `FactoryManager` and `Factory` be `Class` (instead of `T`), perhaps constrained to be something which extends `new (...args: any[]) => any`.